File: app/src/main/java/com/pgaofeng/basemvp/main/view/MainActivity.java
package com.pgaofeng.basemvp.main.view;

import android.os.Handler;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.pgaofeng.basemvp.R;
import com.pgaofeng.basemvp.main.contract.MainContract;
import com.pgaofeng.basemvp.main.presenter.MainPresenter;
import com.pgaofeng.common.base.BaseActivity;


/**
 * @author gaofengpeng
 * @date 2019/3/25
 * @description :View界面，只负责显示。逻辑处理由Presenter负责。
 */
public class MainActivity extends BaseActivity<MainPresenter> implements MainContract.View {

    TextView mTextView;
    Button mButton;


    @Override
    protected int getContentView() {
        return R.layout.activity_main;
    }

    @Override
    protected void initView() {
        mTextView = findViewById(R.id.tv_myText);
        mButton = findViewById(R.id.btn_update);
        mButton.setOnClickListener(v -> {
            mPresenter.updateTextViewText();
        });
    }

    @Override
    protected MainPresenter createPresenter() {
        return new MainPresenter(this,new Handler());
    }

    @Override
    public void updateText(String text) {
        mTextView.setText(text);
    }

    @Override
    public void showToast(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/study/bspatch/main/MainActivity.kt
package com.study.bspatch.main

import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.study.bspatch.PatchUtils
import com.study.bspatch.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

/**
 * @author gaofengpeng
 * @date 2021/9/3
 */
class MainActivity : AppCompatActivity() {

    private lateinit var mTvText: TextView
    private lateinit var mButton: Button
    private lateinit var mButtonToast: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mTvText = findViewById(R.id.tv_myText)
        mButton = findViewById(R.id.btn_update)
        mButtonToast = findViewById(R.id.btn_version)
        init()
    }

    private fun init() {
        // 用于创建cache目录，方便将patch文件放在Android/data/com.study.bspatch/cache目录
        try {
            File(externalCacheDir!!.absolutePath)
        } catch (ignore: Exception) {
        }
        // 可以修改这里文本，然后打包两个不同的安装包
        mButtonToast.setOnClickListener {
            Toast.makeText(this, "我现在是新版本了哦哈哈哈哈哈", Toast.LENGTH_SHORT).show()
        }

        mButton.setOnClickListener {
            val patchFile = File("$externalCacheDir/patch")
            if (!patchFile.exists()) {
                mTvText.text = "未找到分包文件patch，请将其放在{$externalCacheDir}目录下"
                return@setOnClickListener
            }

            lifecycleScope.launch {
                mTvText.text = "开始合并安装包..."
                val last = System.currentTimeMillis()
                val isSuccess = withContext(Dispatchers.IO) {
                    PatchUtils.bsPatch(
                        newFile = "$externalCacheDir/output.apk",
                        oldFile = applicationInfo.sourceDir,
                        patch = "$externalCacheDir/patch"
                    )
                }
                if (isSuccess) {
                    val cost = System.currentTimeMillis() - last
                    mTvText.text = "合并完成，耗时: $cost ms"
                } else {
                    mTvText.text = "合并失败"
                }
            }
        }
    }
}